add validate-config action and README

zkamvar opened the pull request:
This will fix #9 by implementing a validation action for config files
that additionally comments on the pull request the results of the
validation.

annakrystalli reviewed (changes requested):
This is awesome!! 

I made some comments and suggestions for changes directly in the PR.

One last thing I forgot to mention is we also maintain details of available workflows in `hubDocs`.  https://hubverse.io/en/latest/quickstart-hub-admin/continuous-integration.html#available-hubverse-github-actions

Any chance you can add an entry there too?  (might be worth a custom PR template for this repo to remind us every time!)

zkamvar commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,65 @@
+      - '!**README**'
+
+jobs:
+  validate-submission:
```
🙈 not my copy-paste showing through!

zkamvar commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,65 @@
+
+      - uses: r-lib/actions/setup-r-dependencies@v2
+        with:
+          packages: any::gt, hubverse-org/hubValidations, any::sessioninfo
```
> Until they are on CRAN but given we now have an r-universe, should we be using that to install from instead of GitHub for the time being?

Oh we can actually set the R-universe repo in the `setup-r` step!
https://github.com/r-lib/actions/tree/v2-branch/setup-r-dependencies#dependencies-in-other-cran-like-repositories

zkamvar replied:
Given that this will affect all of the workflows, I am going to create a separate PR for it after this goes through.

zkamvar commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,65 @@
+        env:
+          PR_NUMBER: ${{ github.event.number }}
+        run: |
+          library("hubValidations")
```
🙈 copy paste error

zkamvar commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,65 @@
+        run: |
+          library("hubValidations")
+          v <- hubAdmin::validate_hub_config(
+            hub_path = "${{ github.workspace }}"
```
Ah good point! I can certainly do that

zkamvar replied:
done!

zkamvar commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,65 @@
+            # write output to HTML
+            tbl <- hubAdmin::view_config_val_errors(v)
+            writeLines("## :x: Invalid Configuration\n", "${{ github.workspace }}/diff.md")
+            cat("\nThe schema in `hub-config/` appear to be invalid. Details about the exact locations of the errors can be found in the table below.\n", file = "${{ github.workspace }}/diff.md", append = TRUE)
```
Thank you! I was unsure of the correct terminology for this message

annakrystalli commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,67 @@
+        env:
+          PR_NUMBER: ${{ github.event.number }}
+        run: |
+          workspace <- "${{ github.workspace }}"
```
```suggestion
          hub_path <- "${{ github.workspace }}"
```
To be more intuitive?

annakrystalli replied:
In all honesty I was envisaging assigning this to an environment variable like `PR_NUMBER: ${{ github.event.number }}` is above that can be used throughout the step/job as necessary.

e.g is it required again in the comment on PR step or is that sth different?

zkamvar replied:
Ahhhh that makes sense. I will go ahead and set it as an environment variable at the top of the job.

zkamvar replied:
This way, when we port it over to https://github.com/hubverse-org/hubUtils/issues/158, we will simply append the workspace with the correct path.

annakrystalli commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,67 @@
+        uses: carpentries/actions/comment-diff@main
+        with:
+          pr: ${{ github.event.number }}
+          path: ${{ github.workspace }}/diff.md
```
Should this also change if I want to set a custom hub_path?

zkamvar replied:
Yes. I can definitely set this up so that the hub path is set upstream in the job.

zkamvar commented:
`final_this_one_srsly_actually_final.docx`

I added @annakrystalli's suggestions and did the following:

1. updated checkout to v4
2. use the r-universe instead of github for installation
3. improved messaging to give a timestamp and to give a successful message when passing

zkamvar commented:
Also, I have https://github.com/hubverse-org/hubDocs/pull/164 in draft. It's actually ready, but I didn't want to open it until this is merged.

annakrystalli reviewed (commented):
Hehehehe almost! I made a few suggested editing to make it cleaner and more intuitive for someone wanting to override the default `hub_path`.

It did make me think about the workflow `paths` triggers though. These also need to be responsive to a change in `hub_path` otherwise we need to let users know to change them too right?

zkamvar commented on validate-config/validate-config.yaml:
```diff
@@ -0,0 +1,80 @@
+    branches: main
+    paths:
+      - 'hub-config/**'
+      - '!**README**'
```
🤔 I don't think that's possible. I think the best we can do is to give examples of how to modify it. I will do so in my next push.

zkamvar commented on validate-config/README.md:
````diff
@@ -0,0 +1,45 @@
+
+For hubs and repositories which differ in configuration, workflow dispatch will need to be customised manually in the hub's workflow file **in two places**: `on/pull_request/paths` and the `jobs/validate-hub-config/env/HUB_PATH` environment variable (see example below).
+
+For example, to validate the config of a demo hub included as part of a package you would want make sure the hub path is set to `"inst/demo_hub"`:
+
+```diff
+ on:
+   workflow_dispatch:
+   pull_request:
+     branches: main
+     paths:
+-      - 'hub-config/**'
++      - 'inst/demo_hub/hub-config/**'
+       - '!**README**'
+ 
+ jobs:
+   validate-hub-config:
+     runs-on: ubuntu-latest
+     env:
+       GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
+       PR_NUMBER: ${{ github.event.number }}
+-      HUB_PATH: ${{ github.workspace }}
++      HUB_PATH: ${{ github.workspace }}/inst/demo_hub
+```
````
@annakrystalli, I've modified the README to show the example. Does this seem good to you?

annakrystalli replied:
Perfect!

annakrystalli reviewed (approved):
Brilliant! This looks fab now and very clear instructions. Approved!
